Fix rpc nonce vs transaction count

konradkonrad opened the pull request:
Since
1) the testnet/`morden` doesn't use account nonces starting from 0, 
2) `pyethapp/rpc_client.py::eth_sendTransaction` wants to allow for client side signed transactions
we need to have a stable way to figure out nonces in `rpc_client`.

This PR adds a new (non-spec!) jsonrpc-method `eth_nonce` that is used in `rpc_client`. It also adds a test and expands `test_jsonrpc.py` to be also run against testnet like `initial account nonce` settings.

konradkonrad committed:
Use testnet nonces in test, too

konradkonrad committed:
Stabilize finalizer

Apparently we need the context switch to let all services stop cleanly.

konradkonrad committed:
Add non-spec 'eth_nonce' rpcserver method

Since we want to support client side signing of transactions,
we need a way to figure out the account nonces on client side, too.

Because (at least) the testnet uses initial account nonces != 0, we
can no longer rely on `eth_getTransactionCount`.

This adds a jsonrpc method that returns account nonces for (address,
block_id [optional]) tuples.

konradkonrad committed:
Fixup sender/signing logic and allow empty nonce

Prior to this, it wasn't possible to provide `nonce=None`.

konradkonrad committed:
Add test for non-spec `eth_nonce` rpc method

hackaugusto commented on pyethapp/rpc_client.py:
```diff
@@ -150,7 +150,7 @@ def nonce(self, address):
             address = address.decode('hex')
 
-        res = self.call('eth_getTransactionCount', address_encoder(address), 'pending')
+        res = self.call('eth_nonce', address_encoder(address), 'pending')
```
This will only work with `pyethapp`, isn't that a problem?

konradkonrad replied:
Good point. This is the `rpc_client` of `pyethapp`. I don't know, if anyone considers that a standalone rpc-client. If worst comes to worse, that could be the case.
For `JSONRPCClient.nonce(self)` I wouldn't know how to solve it -- endpoint probing seems overkill. I'll try to add a catch for the missing server method with a reasonable error message. 

For `JSONRPCClient.eth_sendTransaction(self, ...)` `nonce=None` is a valid parameter that get's resolved through the server endpoints (I believe that is how other clients do it).

hackaugusto commented on pyethapp/tests/test_jsonrpc.py:
```diff
@@ -205,7 +209,12 @@ def rpc_request(self, method, *args):
+                app.services[service].stop()
+            except Exception as e:
+                log.DEV(str(e), exc_info=e)
+                pass
```
we might also want to kill the pending greenlets here:

``` python
gevent.killall(task for task in gc.get_objects() if isinstance(task, gevent.Greenlet))
```

konradkonrad replied:
I hadn't had any problems anymore with above finalizer. But it also probably won't hurt to clean up thoroughly.

konradkonrad committed:
PEP8

konradkonrad committed:
Improve finalizer

konradkonrad committed:
Add failure message for cross client application

`pyethapp`'s jsonrpc has a few non-spec methods. For `eth_nonce` we
catch the `Method not found` error and raise an Exception with a more
specific error message.

konradkonrad committed:
Some documentation fixes

konradkonrad committed:
PEP8

konradkonrad committed:
Missing dependency

`os` is missing at this point, but branch `use_rpc_tests` should
make this obsolete.
